### ISSUE TITLE
Unable to upload the firmware to the Cheetah v1.1 board 

### ISSUE THREAD

**Issue by EnnajimiYassine** [FYSETC/Cheetah#2]
I'm unable to upload the firmware to my cheetah v1.1 , I tried to upload with a mac and a raspberry pi but I get the same error. Am I missing something ? :
stm32flash 0.5
http://stm32flash.sourceforge.net/
Using Parser : Intel HEX
Interface serial_posix: 57600 8E1
GPIO sequence start
setting port signal rts to 1... OK
delay 100000 us
setting port signal dtr to 0... OK
GPIO sequence end
Failed to init device

Thank you for your help

**Comment by puccac** [comment]
Programs required for firmware upload in Windows

Git CMD
https://git-scm.com/downloads
Download for Windows

stm32flash
https://sourceforge.net/projects/stm32flash/
Download stm32flash-0.5-win64.zip

How to upload firmware

3d Turn off the printer, connect the computer to the board with a USB cable, and check the COM port.

open git cmd
cd Downloads \ stm32flash-0.5-win64 <-stm32flash path

Read information
stm32flash -b460800 COM3 <-change after checking the port

Write firmware
stm32flash -R -b460800 -w firmware.hex -v -i rts, -dtr, dtr COM3 <-change after checking the port

**Comment by empakoso** [comment]
OctoPrint will also work.  I just used plug-in Firmware Updater and STM32 drivers (see the Firmware Updater GitHub page to perform STM32 drivers using RPi SSH).  Settings:
Flash method (select) "stm32flash (STM32 built-in bc..." 
Set the path to stm32flash (type) "/usr/bin/stm32flash" 
Test (click result) "The path is valid"

I was able to connect and flash successful using the hex file

**Comment by db414813** [comment]
I have same issue as OP. I have the required software but get same error as OP. Any resolution to this? I really don't want to have to flash an rpi with octoprint when I should not have to.

**Comment by db414813** [comment]
I gave up and followed the rpi octopi route. Flash was successful per octopi. I will report back any issues. I am still not sure why my linux desktop could not flash the firmware to the cheetah board.

**Comment by Magoo0876** [comment]
Connect the board to the octopi. 
Open a linux command line with:
    ssh pi@octopi.local
enter the password: raspberry 
or if changed what you have chosen
check your usb connection
   dmesg |grep usb

there should  be something like that:
[ 1078.737260] usb 1-1.4: new full-speed USB device number 6 using dwc_otg
[ 1078.870413] usb 1-1.4: New USB device found, idVendor=1a86, idProduct=7523, bcdDevice= 2.63
[ 1078.870428] usb 1-1.4: New USB device strings: Mfr=0, Product=2, SerialNumber=0
[ 1078.870438] usb 1-1.4: Product: USB2.0-Serial
[ 1078.874440] **usb 1-1.4: ch341-uart converter now attached to ttyUSB0**

You should see a correct connection. Procced with flashing. 
If not update the raspberry pi.
type:
    sudo apt full-upgrade

   sudo reboot
#type:
    ls /dev/serial/by-id/

Enter the id in the next command
 
    stm32flash -w /home/pi/yourfirmwarefile.bin -v -i rts,-dtr,dtr /dev/serial/by-id/enteryouridhere

hth regards Magoo

**Comment by Jugulaire** [comment]
No, it's not working on any of my machine. Tested on OPI one,Zero Raspberrypi 2,3,4 and finally with my Lenovo T450 and X220? no luck ... 

```bash
┌🤘-🐧jugu@ 💻 W0RKST4TI0N - 🧱 11F5-30EA
└🤘-> sudo stm32flash -w firmware.bin -v -i rts,-dtr,dtr /dev/ttyUSB0 -R 
[sudo] Mot de passe de jugu : 
stm32flash 0.5

http://stm32flash.sourceforge.net/

Using Parser : Raw BINARY
Interface serial_posix: 57600 8E1

GPIO sequence start
 setting port signal rts to 1... OK
 delay 100000 us
 setting port signal dtr to 0... OK
 delay 100000 us
 setting port signal dtr to 1... OK
GPIO sequence end

Failed to init device.
```

**Comment by sixtyfive** [comment]
Same problem. Output of `stm32flash /dev/ttyUSB0`:

```
stm32flash 0.6

http://stm32flash.sourceforge.net/

Interface serial_posix: 57600 8E1
Version      : 0x22
Option 1     : 0x00
Option 2     : 0x00
Device ID    : 0x0414 (STM32F10xxx High-density)
- RAM        : Up to 64KiB  (512b reserved by bootloader)
- Flash      : Up to 512KiB (size first sector: 2x2048)
- Option RAM : 16b
- System RAM : 2KiB
```

output of `stm32flash -w out/klipper.bin -v -i rts,-dtr,dtr /dev/ttyUSB0`:

```
stm32flash 0.6

http://stm32flash.sourceforge.net/

Using Parser : Raw BINARY
Interface serial_posix: 57600 8E1

GPIO sequence start
 setting port signal rts to 1... OK
 delay 100000 us
 setting port signal dtr to 0... OK
 delay 100000 us
 setting port signal dtr to 1... OK
GPIO sequence end

Failed to init device.
```

Please help!

**Comment by Magoo0876** [comment]
try:

**dmesg |grep tty**

and look if ttyUSB0 is the correct device.

pi@mainsailos:~/klipper $ dmesg |grep tty
[    0.000000] Kernel command line: coherent_pool=1M 8250.nr_uarts=0 snd_bcm2835.enable_compat_alsa=0 snd_bcm2835.enable_hdmi=1 bcm2708_fb.fbwidth=720 bcm2708_fb.fbheight=720 bcm2708_fb.fbswap=1 vc_mem.mem_base=0x3ec00000 vc_mem.mem_size=0x40000000  console=ttyS0,115200 console=tty1 root=PARTUUID=1b18721f-02 rootfstype=ext4 elevator=deadline fsck.repair=yes rootwait
[    0.001187] printk: console [tty1] enabled
[    3.290891] 3f201000.serial: ttyAMA0 at MMIO 0x3f201000 (irq = 114, base_baud = 0) is a PL011 rev2
[964733.566065] usb 1-1.5: **ch341-uart converter now attached to ttyUSB0**

If this is ok the set the flash command two times immediately one after one. 
At my board it always works at the second attempt.

Try another usb cable even if the device is correctly connected and visible at the linux os.

**Comment by Magoo0876** [comment]
This is my actual right now tested command for flashing.
Check the /dev/serial/by-id folder if there is a file called like the last part of the command below. If not change it to your USB id.

stm32flash -w /home/pi/klipper/out/klipper.bin -v -i rts,-dtr,dtr /dev/serial/by-id/**usb-1a86_USB2.0-Serial-if00-port0**

**Comment by sixtyfive** [comment]
Hi @Magoo0876, thanks for the quick response!

I'd tried with both `/dev/serial/by-id/usb-1a86_USB2.0-Serial-if00-port0`, as well as `/dev/ttyUSB0` (making sure it's the only / correct one) before; no difference. I had also been playing with the GPIO sequence, as various places (Wiki, Issues, forums) give different recommendations (so far have seen: `rts,-dtr,dtr`, `rts,-dtr`, `rts,dtr`). I've also tried: multiple times one after the other manually; multiple times in a `for` loop; a script polling for `/dev/serial/by-id/usb-1a86_USB2.0-Serial-if00-port0` and then sending the command right after the board is plugged in. All to no avail.

What I've not tried is a different cable ... I'll go out and try to find another one today!

**Comment by sixtyfive** [comment]
Okay, so I went out today and searched shops until I found a mini USB cable. It cost 11€ for 75cm and is "premium shielded" (whatever that's supposed to mean, only some marketing person will know). Unfortunately it doesn't change a thing. Still `Failed to init device.` Now having two cables, I cut the +5V on one of them to get rid of the backfeed problem while powered from +12V. Again, no change (wouldn't have expexted one). Always trying at least twice in a row by the way.

Unless you have any more ideas, I'll go out and get an stlinkv2 and try via the SWD header :-/

**Comment by Magoo0876** [comment]
So that's not nice. What you can still try is to update your os. 

If you are on an raspberry or Debian base type:

**sudo apt update && apt full-upgrade**
and for the firmware of the pi

**sudo rpi-update**

In the past there was a problem with the ch340/341 driver which I also ran into it.
If this still does not do the job, flash the stock firmware via sd card to the cheetah board to look if the board is ok.  
Then copy the Klipper.bin file to the sd and try it this way.
I also attached my firmware. To use it you have to rename it!


[klipper.txt](https://github.com/FYSETC/FYSETC-Cheetah/files/6940966/klipper.txt)

Good luck.

**Comment by sixtyfive** [comment]
Okay, so, it just worked. I didn't try another time; bit scared to break something. The command that eventually made it work was:

```
stm32flash -w out/klipper.bin -v -i rts /dev/serial/by-id/usb-1a86_USB2.0-Serial-if00-port0
```

For the record, my board is version 1.1b and my laptop is running Manjaro Linux (I tend to keep all packages up-to-date quite pedantically). Kernel version at the time of this writing was 5.10.53.
